storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala

```diff
@@ -0,0 +1,364 @@
+        case (true, true)   => CharClassPattern(this.charSet & thatCharClass.charSet, true)
+        case (true, false)  => CharClassPattern(this.charSet -- thatCharClass.charSet, true)
+        case (false, true)  => CharClassPattern(thatCharClass.charSet -- this.charSet, true)
+        case (false, false) => CharClassPattern(this.charSet | thatCharClass.charSet)
+      }
+    case _ => multiply(1) | that
+  }
+
+  /**
+    * Return a RegexPattern that can match any string
+    *	that both self and other can match.
+    *
+    * ¬A AND ¬B = ¬(A OR B)
+    * ¬A AND B = B - A
+    * A AND ¬B = A - B
+    * A AND B
+    */
+  override def intersection(that: RegexPattern): RegexPattern = that match {
+    case thatCharClass: CharClassPattern =>
+      (this.negateCharSet, thatCharClass.negateCharSet) match {
+        case (true, true)   => CharClassPattern(this.charSet | thatCharClass.charSet, true)
+        case (true, false)  => CharClassPattern(thatCharClass.charSet -- this.charSet)
+        case (false, true)  => CharClassPattern(this.charSet -- thatCharClass.charSet)
+        case (false, false) => CharClassPattern(this.charSet & thatCharClass.charSet)
+      }
+    case _ => multiply(1) & that
+  }
+}
+
+/**
+  * A min and a max. The vast majority of characters in regular
+  * expressions occur without a specific multiplier, which is implicitly
+  * equivalent to a min of 1 and a max of 1, but many more have explicit
+  * multipliers like "*" (min = 0, max = inf) and so on.
+  * Although it seems odd and can lead to some confusing edge cases, we do
+  * also permit a max of 0 (iff min is 0 too). This allows the multiplier
+  * "zero" to exist, which actually are quite useful in their own special way.
+  */
+final case class Multiplier(min: Option[Int], max: Option[Int]) {
+  def this(min: Int, max: Int) = this(Some(min), Some(max))
+  def this(multiplier: Int) = this(Some(multiplier), Some(multiplier))
+
+  require(
+    (min, max) match {
+      case (Some(x), Some(y)) => (0 <= x) && (x <= y)
+      case (Some(x), None)    => 0 <= x //min valid, max infinite
+      case (None, Some(_))    => false //min is infinite
+      case (None, None)       => true
+    },
+    "Invalid multiplier bounds"
+  )
+
+  /**
+    * We need some math operation on Option[Int], assuming that None == infinite
+    */
+  implicit class OptionIntMath(opt: Option[Int]) {
+    def +(that: Option[Int]): Option[Int] = (opt, that) match {
+      case (Some(_), None)    => None
+      case (None, Some(_))    => None
+      case (None, None)       => None
+      case (Some(x), Some(y)) => Some(x + y)
+    }
+
+    /**
+      * Subtract another bound from this one.
+      * Caution: this operation is not meaningful for all bounds.
+      */
+    def -(that: Option[Int]): Option[Int] = (opt, that) match {
+      case (Some(_), None) =>
+        throw new IllegalArgumentException("Can't substract infinity") //something - infinity => invalid operation
```

Comment:
Could we replace exception with something like `Either[Error,Result]`?

Reply:
Again - this error can't appear using API, since Multiplier operations are not available outside of a package.